runtimes/runtimes/chat/encryptedChat.ts

```diff
@@ -0,0 +1,156 @@
+import { jwtDecrypt, EncryptJWT, JWTPayload } from 'jose'
+import { Connection } from 'vscode-languageserver'
+import {
+    ChatParams,
+    ChatResult,
+    EndChatParams,
+    FeedbackParams,
+    FollowUpClickParams,
+    InfoLinkClickParams,
+    InsertToCursorPositionParams,
+    LinkClickParams,
+    NotificationHandler,
+    QuickActionParams,
+    QuickActionResult,
+    RequestHandler,
+    SourceLinkClickParams,
+    TabAddParams,
+    TabChangeParams,
+    TabRemoveParams,
+    chatRequestType,
+    feedbackNotificationType,
+    readyNotificationType,
+    tabAddNotificationType,
+    tabChangeNotificationType,
+    tabRemoveNotificationType,
+    insertToCursorPositionNotificationType,
+    linkClickNotificationType,
+    infoLinkClickNotificationType,
+    sourceLinkClickNotificationType,
+    followUpClickNotificationType,
+    endChatRequestType,
+    CancellationToken,
+    EncryptedChatParams,
+    EncryptedQuickActionParams,
+    quickActionRequestType,
+} from '../../protocol'
+import { Chat } from '../../server-interface'
+import { CredentialsEncoding } from '../auth/standalone/encryption'
+
+export class EncryptedChat implements Chat {
+    private key: Buffer
+    private encoding: CredentialsEncoding | undefined
+
+    constructor(
+        private readonly connection: Connection,
+        key: string,
+        encoding?: CredentialsEncoding
+    ) {
+        this.key = Buffer.from(key, 'base64')
+        this.encoding = encoding
+    }
+
+    public onChatPrompt(handler: RequestHandler<ChatParams, ChatResult | null | undefined, ChatResult>) {
+        this.connection.onRequest(chatRequestType, async (request: ChatParams | EncryptedChatParams) => {
+            request = request as EncryptedChatParams
+            // decrypt the request params
+            let decryptedRequest = (await this.decodeRequest(request)) as ChatParams
+
+            // make sure we don't lose the partial result token
+            if (request.partialResultToken) {
+                decryptedRequest.partialResultToken = request.partialResultToken
+            }
+
+            // call the handler with plaintext
+            const response = (await handler(decryptedRequest, CancellationToken.None)) as ChatResult
+            // encrypt the response
+            const encryptedResponse = await this.encryptRequest(response as JWTPayload)
+            // send it back
+            return encryptedResponse
+        })
+    }
+
+    public onQuickAction(handler: RequestHandler<QuickActionParams, QuickActionResult, void>) {
+        this.connection.onRequest(
```

Comment:
nit: code duplication in handling encrypted chatPrompt and quickAction could be reduced

Reply:
I feel this whole class can be replaced with a generated type and a generic implementation of `onRequest` that automatically converts the intended `*RequestType` to the corresponding `Encrypted*RequestType` without typing it out explicitly.

Reply:
I spent some time trying to use generics and define a single function but it was taking long as it was not as straightforward as I expected so I left that for later